File: HPWorkpath_v1.6.3/Samples/ExampleAPIServices/source/StatisticsSample/src/com/hp/workpath/sample/statisticsample/task/JobInfoTask.java
// Copyright 2018 HP Development Company, L.P.
// SPDX-License-Identifier: MIT
package com.hp.workpath.sample.statisticsample.task;

import android.content.Context;
import android.os.Handler;
import android.os.Looper;

import com.hp.workpath.api.Result;
import com.hp.workpath.api.statistics.StatisticsJobData;
import com.hp.workpath.api.statistics.StatisticsService;
import com.hp.workpath.sample.statisticsample.Logger;
import com.hp.workpath.sample.statisticsample.fragment.ResponseInterface;

import java.lang.ref.WeakReference;
import java.util.ArrayList;
import java.util.List;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;

public class JobInfoTask {

    private WeakReference<Context> mContextRef;
    private ResponseInterface mResponseInterface;
    private Result mResult;
    private ExecutorService executor = Executors.newSingleThreadExecutor();
    private Handler handler = new Handler(Looper.getMainLooper());

    public JobInfoTask(Context context, ResponseInterface responseInterface) {
        this.mContextRef = new WeakReference<>(context);
        this.mResponseInterface = responseInterface;
        this.mResult = new Result();
    }

    public void taskExecute(Integer... args) {
        try {
            executor.execute(new Runnable() {
                @Override
                public void run() {
                    Integer jobSequence = null;
                    if (args.length == 1) {
                        jobSequence = args[0];
                    }

                    try {
                        if (jobSequence == null) {
                            onPostExecute(StatisticsService.getAllJobsList(mContextRef.get(), mResult));
                        } else {
                            List<StatisticsJobData> jobDataList = new ArrayList<>();
                            jobDataList.add(StatisticsService.getJobInfoByJobSequence(mContextRef.get(), jobSequence, mResult));
                            onPostExecute(jobDataList);
                        }
                    } catch (Throwable t) {
                        Logger.showResult(null, "StatisticsService.getJobInfoByJobSequence " + t.getMessage());
                        executor.shutdown();
                        onPostExecute(null);
                    }

                }
            });
        } catch (Exception e) {
            Logger.showResult(null, "StatisticsService.getJobInfoByJobSequence " + e.getMessage());
            onPostExecute(null);
            executor.shutdown();
        }
    }

    private void onPostExecute(List<StatisticsJobData> jobDataList) {
        handler.post(new Runnable() {
            @Override
            public void run() {
                if (mResult.getCode() == Result.RESULT_OK && jobDataList != null) {
                    mResponseInterface.success(jobDataList);
                } else {
                    mResponseInterface.failure("StatisticsService.getJobInfoByJobSequence", mResult);
                }
            }
        });
    }
}

File: HPWorkpath_v1.6.3/Samples_Kotlin/ExampleAPIServices/source/StatisticsSample/src/com/hp/workpath/sample/statisticsample/task/JobInfoTask.kt
// Copyright 2018 HP Development Company, L.P.
// SPDX-License-Identifier: MIT
package com.hp.workpath.sample.statisticsample.task

import android.content.Context
import com.hp.workpath.api.Result
import com.hp.workpath.api.statistics.StatisticsJobData
import com.hp.workpath.api.statistics.StatisticsService
import com.hp.workpath.sample.statisticsample.fragment.ResponseInterface
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import java.lang.ref.WeakReference

class JobInfoTask(context: Context, responseInterface: ResponseInterface) {
    private val mContextRef: WeakReference<Context> = WeakReference(context)
    private val mResponseInterface: ResponseInterface = responseInterface
    private var mThrowable: Throwable? = null
    private val mResult: Result = Result()

    suspend fun execute(jobSequence: Int? = null) {
        try {
            mContextRef.get()?.run {
                return if (jobSequence == null) {
                    onPostExecute(StatisticsService.getAllJobsList(this, mResult))
                } else {
                    val jobDataList: MutableList<StatisticsJobData> = ArrayList()
                    jobDataList.add(StatisticsService.getJobInfoByJobSequence(this, jobSequence, mResult))
                    onPostExecute(jobDataList)
                }
            }
        } catch (t: Throwable) {
            mThrowable = t
        }
        onPostExecute(null)
    }

    private suspend fun onPostExecute(jobDataList: List<StatisticsJobData>?) {
        withContext(Dispatchers.Main) {
            if (mResult.code == Result.RESULT_OK && jobDataList != null) {
                mResponseInterface.success(jobDataList)
            } else {
                mThrowable?.run {
                    mResponseInterface.failure("StatisticsService.getJobInfoByJobSequence $message", null)
                } ?: run {
                    mResponseInterface.failure("StatisticsService.getJobInfoByJobSequence", mResult)
                }
            }
        }
    }
}